### ISSUE TITLE
[Bug]: Browsers take several minutes to open a window when using "IVPN" integrated VPN/firewall

### ISSUE THREAD

**Issue by C0deViewer** [flatpak/flatpak#5367]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.14.4

### What Linux distribution are you using?

Fedora Linux

### Linux distribution version

37

### What architecture are you using?

x86_64

### How to reproduce

Prerequisitions: a Linux PC system with Flatpak installed

1. Install a app from Flathub
2. Try to open the app
3. See if you get immediately a window after starting the installed app

### Expected Behavior

The window opens directly without having to wait around 5 minutes until a window of the app opens.

### Actual Behavior

It takes around around 5 minutes until app starts to show a window. See the video for the demonstration down below (if visible and able to start). Depending on the app (e.g. Firefox) it won't even open a window at all and shuts its process down after around a minute.

[Issue - Flatpak app needs around 5 minutes to open window.webm](https://user-images.githubusercontent.com/126986902/227531415-1c5fe3d1-f843-4fbd-ae83-e226f563705b.webm)

### Additional Information

The problem happens on Wayland and X11, so no difference at this part. I also saw already a bug ([https://github.com/flatpak/flatpak/issues/4211](url)), which described my problem. The reason why I still make a new issue for that is 1. it's already closed, 2. it's from 2021 (almost 2 years old), 3. the solution of the mentioned issue didn't changed something and 4. it looks like that the bug has been resolved already. I believe that the bug here is different to the mentioned issue*. I already talked about this issue on Fedora's own forum where I believe that it is somehow connected with apps which are using hardware acceleration (like Firefox/Librewolf). At the end, I'm not even sure if hardware acceleration has to do something with that or not. So I hope that I can find here help. Thanks in advance for any try to help!

*_This is my very first issue on GitHub generally, so I'm sorry if this issue doesn't provide enough useful informations now. And I'm also sorry if this issue doesn't met the requirements (regarding the part of "if the issue already exists in the issue tracker") completely. It wasn't a easy decision there how to handle this 100% correctly. Thanks for understanding!_

**Comment by hyuri** [comment]
Same here. After startup, Chromium and other apps takes at least 5 minutes to start. Smaller, simpler apps take at least a minute. Sometimes the app doesn't even open (I noticed its processes die) and I have to click to open again.

After that very slow first run, Chromium and other apps start faster (30 seconds, if I had to guess), but that's still way way slower than native packages and Snaps (Snaps take about 2 seconds to open).

- Pop!_OS 22.04
- Flatpak 1.14.4

**Comment by C0deViewer** [comment]
@smcv My issue is related with this [issue](https://github.com/ivpn/desktop-app/issues/259) of an VPN app with an integrated firewall option (I'm testing such apps currently). The firewall tries to get every connection through the VPN tunnel. For some reason, the firewall creates a IP(v6) traffic burst if someone starts a Flatpak app which needs to connect somewhere right at start up (for example Firefox, Thunderbird, etc.). The IP burst blocks programs to open their windows like they should (instantly after starting them).

It's unsure if not unlikely that the Flatpak project is able to do something here regarding this problem. I would close this issue if that's true then.

**Comment by smcv** [comment]
We *definitely* can't do anything about this without more information, most importantly whatever log messages appear when this happens. Sorry, I am not going to watch a 5 minute video of an app not starting in the hope that it contains information; please copy/paste text if available.

We probably can't fix a VPN app with an integrated firewall; if this failure is specific to this VPN app, then there is probably something unusual it is doing, which might conflict with how Flatpak's sandbox gets set up. If it has a mode where it can be asked to provide a more verbose log about what it's doing, that might provide helpful information.

For apps that have the `--share=network` permission (like Firefox), Flatpak doesn't do anything particularly special about the network namespace: the same network devices that are available to the host are available to the sandbox. (Apps with `--unshare=network` can't access the network at all.)

For apps that have the `--share=network` permission, Flatpak tries to make the host system's DNS resolution arrangements available inside the sandbox (meaning `/etc/resolv.conf`, and the `systemd-resolvd` socket if you are using that).

It might be helpful to run one of the affected apps with a command like `flatpak -v -v --command=bash --devel org.mozilla.Firefox`. That will:

* give you an interactive shell inside the sandbox, from which you can run the actual app, possibly under strace or similar (`--command=bash`);
* use a "SDK" runtime containing more debugging tools (`--devel`);
* use a somewhat more permissive seccomp filter that is less likely to break debugging tools (`--devel`);
* log verbose messages about what Flatpak is doing when it sets up the sandbox (`-v -v`).

**Comment by smcv** [comment]
Another useful thing to try would be to use the simplest app you can, to find out whether this is specific to a particular category of apps (like "all web browsers" or "apps that make a network connection during startup") or whether it is literally all apps. For example, https://flathub.org/apps/org.gnome.Recipes is a much smaller/simpler app than Firefox (but does require networking), and https://flathub.org/apps/org.gnome.font-viewer is a small/simple app that *doesn't* require networking.

**Comment by smcv** [comment]
@hyuri:
> I've just read in multiple threads that removing xdg-desktop-portal / xdg-desktop-portal-gnome / xdg-desktop-portal-kde fixed the issue for a lot of people. Not a solution since that breaks apps, of course. But the issue seems to have something to do with at least one of these.

If you are not using this IVPN thing, and you don't find an obvious solution when you check the logs, then it is probably better to report what you are seeing as a *separate* issue, and be as specific as possible (error messages, etc.) to avoid unrelated issues being bundled together with yours. If it turns out to be an aspect of the same thing that @C0deViewer is seeing, then it's easy to close the second issue as a duplicate, but if it's a different root cause, keeping them both together on #5367 will get really confusing - and the more confusing a bug is, the longer it will take anyone to solve it. So it's good for everyone, including you, if you make it as straightforward as possible to understand what is happening.

If you have evidence that xdg-desktop-portal is somehow involved in what you are seeing, then definitely open that as a separate issue.

The same debugging steps I recommended for @C0deViewer (checking the system log, `flatpak -v -v`, etc.) should be equally helpful for you.

Removing xdg-desktop-portal should never be necessary, and will break most Flatpak apps.

Removing the desktop-environment-specific backend for desktops that you don't use (for example removing x-d-p-gnome if you are not a GNOME user, or x-d-p-kde if you are not a KDE user) can be helpful if you are encountering the failure mode where one of the backends fails to start because it was run in the wrong desktop environment, but then the failure to start doesn't get properly signalled back to the non-desktop-environment-specific portal interface. Looking at the system log (systemd Journal or syslog) would tell you whether this is happening.

**Comment by C0deViewer** [comment]
@smcv Thanks for the suggestions. I can say so far that I already tried looking on my system logs (`journalctl`) and it didn't provided really useful informations regarding one of the affected apps ([LibreWolf](https://librewolf.net/), affected by this problem like Firefox) which I used for this test now. The only thing it showed regarding LibreWolf was this string: 
`Jun 09 19:48:50 pc systemd[3004]: Started app-flatpak-io.gitlab.librewolf\x2dcommunity-3627.scope.`
After that, nothing came up further. No potential warnings, errors, etc.

Running Librewolf via Flatpak shell tells me that it interacts with `resolv.conf` at least (and with that also with `systemd-resolved` probably like it was mentioned on the issue on IVPN's GitHub page). Looks like this in terminal:

> F:     --symlink
> F:     /run/host/monitor/resolv.conf
> F:     /etc/resolv.conf

I can also confirm that the bug **doesn't** happen with apps like [LibreOffice](https://flathub.org/apps/org.libreoffice.LibreOffice), [Loupe](https://flathub.org/apps/org.gnome.Loupe) or [Resonance](https://flathub.org/apps/io.github.nate_xyz.Resonance) for example.

The bug only happens if I activate the firewall function from IVPN. Everything works normal like it should if the firewall is deactivated.

PS: Sorry if I don't provide the informations you're asking for maybe. I try my best to provide them but I'm also still new at troubleshooting something like this, especially if it's connected with Flatpak directly.
